### PR TITLE
Add rewrite prototype and couple of fixes

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 1]
```diff
@@ -204,7 +204,7 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
   def DefDef_apply(symbol: Symbol, rhsFn: List[Type] => List[List[Term]] => Option[Term])(given Context): DefDef =
     withDefaultPos(tpd.polyDefDef(symbol.asTerm, tparams => vparamss => rhsFn(tparams)(vparamss).getOrElse(tpd.EmptyTree)))
 
-  def DefDef_copy(original: DefDef)(name: String, typeParams: List[TypeDef], paramss: List[List[ValDef]], tpt: TypeTree, rhs: Option[Term])(given Context): DefDef =
+  def DefDef_copy(original: Tree)(name: String, typeParams: List[TypeDef], paramss: List[List[ValDef]], tpt: TypeTree, rhs: Option[Term])(given Context): DefDef =
     tpd.cpy.DefDef(original)(name.toTermName, typeParams, paramss, tpt, rhs.getOrElse(tpd.EmptyTree))
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: This change seems to break consistency with other types of trees.

File: library/src/scala/quoted/Expr.scala [hunk 2]
```diff
@@ -190,6 +190,11 @@ package internal {
      *  May contain references to code defined outside this TastyTreeExpr instance.
      */
     final class TastyTreeExpr[Tree](val tree: Tree, val scopeId: Int) extends Expr[Any] {
+      override def equals(that: Any): Boolean = that match {
+        case that: TastyTreeExpr[_] => tree == that.tree && scopeId == that.scopeId
+        case _ => false
+      }
```
Comment (liufengyun) [inline library/src/scala/quoted/Expr.scala]: Maybe add a comment about what we can expect if two Expr are equal. For example, why referential equality is not good enough?

File: library/src/scala/tasty/reflect/TreeOps.scala [hunk 3]
```diff
@@ -97,7 +97,7 @@ trait TreeOps extends Core {
   object DefDef {
     def apply(symbol: Symbol, rhsFn: List[Type] => List[List[Term]] => Option[Term])(given ctx: Context): DefDef =
       internal.DefDef_apply(symbol, rhsFn)
-    def copy(original: DefDef)(name: String, typeParams: List[TypeDef], paramss: List[List[ValDef]], tpt: TypeTree, rhs: Option[Term])(given ctx: Context): DefDef =
+    def copy(original: Tree)(name: String, typeParams: List[TypeDef], paramss: List[List[ValDef]], tpt: TypeTree, rhs: Option[Term])(given ctx: Context): DefDef =
```
Comment (liufengyun) [inline library/src/scala/tasty/reflect/TreeOps.scala]: Same question about consistency.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Those should also change. It should be as in the dotty tree copier. I will adapt the missing ones.

File: library/src/scala/tasty/reflect/TreeOps.scala (hunk 3 above)
Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/TreeOps.scala]: Same answer

File: library/src/scala/quoted/Expr.scala (hunk 2 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Expr.scala]: Added a comment

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: Changed them

**Pull request merged**